core/src/main/java/org/apache/carbondata/core/indexstore/DataMap.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.carbondata.core.indexstore;
+
+import java.util.Iterator;
+
+import org.apache.carbondata.core.events.EventListener;
+import org.apache.carbondata.core.scan.filter.resolver.FilterResolverIntf;
+
+/**
+ * Interface for adding and retrieving index data.
+ */
+public interface DataMap<T> extends EventListener {
+
+  /**
+   * Give the writer to write the data.
+   *
+   * @return
+   */
+  DataMapWriter<T> getWriter();
+
+  /**
+   * It is called to load the data map to memory or to initialize it.
+   */
+  DataMapInitializer getInitializer();
```

Comment:
Can we put `init` func in this interface directly?
`DataMap<T>` is the abstraction of the storage, so it can be initialized and queried